modules/compiler/spirv-ll/source/builder_debug_info.cpp

```diff
@@ -15,29 +15,2444 @@
+        if (auto checksum_kind =
+                llvm::DIFile::getChecksumKind(checksum_kind_str)) {
+          size_t checksum_end_pos = checksum_str.find_if_not(llvm::isHexDigit);
+          checksum.emplace(checksum_kind.value(),
+                           checksum_str.substr(0, checksum_end_pos));
+        }
+      }
+    }
+  }
+
+  // It doesn't matter which DIBuilder we use to create a DIFile, as they
+  // exist independently from the CompileUnit hierarchy.
+  return getDefaultDIBuilder().createFile(fileName, fileDir, checksum);
+}
+
+//===----------------------------------------------------------------------===//
+// 4.3 Type Instructions
+//===----------------------------------------------------------------------===//
+
+class DebugTypeBasic : public OpExtInst {
+ public:
+  DebugTypeBasic(OpCode const &other) : OpExtInst(other) {}
+  spv::Id Name() const { return getOpExtInstOperand(0); }
+  spv::Id Size() const { return getOpExtInstOperand(1); }
+  uint32_t Encoding() const { return getOpExtInstOperand(2); }
+};
+
+template <>
+llvm::Expected<llvm::MDNode *> DebugInfoBuilder::translate(
+    const DebugTypeBasic *op) {
+  std::optional<std::string> name = module.getDebugString(op->Name());
+  if (!name) {
+    return makeStringError(
+        "Could not find OpString 'Name' for DebugTypeBasic " +
+        getIDAsStr(op->IdResult(), &module));
+  }
+
+  auto encoding_iter = DebugEncodingMap.find(op->Encoding());
+  if (encoding_iter == DebugEncodingMap.end() || encoding_iter->second == 0) {
+    return getDIBuilder(op).createUnspecifiedType(*name);
+  }
+  llvm::dwarf::TypeKind encoding = encoding_iter->second;
+
+  auto size_or_error = getConstantIntValue(op->Size());
+  if (auto err = size_or_error.takeError()) {
+    return err;
+  }
+  // Without a size, we can't create a type.
+  if (!size_or_error.get()) {
+    return nullptr;
+  }
+  uint64_t size = *size_or_error.get();
+
+  return getDIBuilder(op).createBasicType(*name, size, encoding);
+}
+
+class DebugTypePointer : public OpExtInst {
+ public:
+  DebugTypePointer(OpCode const &other) : OpExtInst(other) {}
+  spv::Id BaseType() const { return getOpExtInstOperand(0); }
+  uint32_t StorageClass() const {
+    return static_cast<spv::StorageClass>(getOpExtInstOperand(1));
+  }
+  uint32_t Flags() const { return getOpExtInstOperand(2); }
+};
+
+template <>
+llvm::Expected<llvm::MDNode *> DebugInfoBuilder::translate(
+    const DebugTypePointer *op) {
+  auto base_ty = translateDebugInst<llvm::DIType>(op->BaseType());
+  if (auto err = base_ty.takeError()) {
+    return err;
+  }
+  unsigned addrspace = 0;
+  auto addrspace_or_error =
+      module.translateStorageClassToAddrSpace(op->StorageClass());
+  if (auto err = addrspace_or_error.takeError()) {
+    // Silently consume this error. We know that llvm-spirv will use ~0 to
+    // represent "no address space", despite this being invalid SPIR-V. We
+    // translate it to address space 0.
+    addrspace = 0;
```

Comment:
Does this address space come from Clang? LLVM IR doesn't have the concept of "no address space".

Reply:
You're right about LLVM.

What `llvm-spirv` does is check whether the `llvm::DIDerivedType` has an address space (which is a `std::optional<unsigned>`). So I suspect that during code generation, clang might omit the address space on things like `int *ptr`? I can double check that. I also just noticed that we don't necessarily have to set it to `0` - we can just forward on the `std::optional`.

Reply:
[`DIBuilder::getPointerType`](https://llvm.org/doxygen/classllvm_1_1DIBuilder.html#a8f163bf8a14c1c7823f6d8fce7fb5ac7) takes the same `std::optional<unsigned>` that the original `DerivedType` had. I think we should preserve that, I think we shouldn't use any particular address space as the default address space, and we should translate `~0U` back to `std::nullopt`.